src/field_theory/krull_topology.lean

```diff
@@ -287,3 +289,233 @@ begin
+end
+
+lemma ultrafilter.glued_generators_of_pushforwards_alg_hom_surjective (h_int :
+  algebra.is_integral K L) (f : ultrafilter (L →ₐ[K] L)) :
+  function.surjective (ultrafilter.glued_generators_of_pushforwards_alg_hom f h_int) :=
+begin
+  intro y,
+  let p := minpoly K y,
+  let S := p.root_set L,
+  let σ := ultrafilter.glued_generators_of_pushforwards_alg_hom f h_int,
+  have hp : p ≠ 0 := (minpoly.monic (h_int y)).ne_zero,
+  have hσSS : σ '' S ⊆ S,
+  { rintros x ⟨a, ha, rfl⟩,
+    rw polynomial.mem_root_set hp at ha,
+    rw [polynomial.mem_root_set hp, polynomial.aeval_alg_hom_apply, ha, σ.map_zero] },
+  have h3 : σ '' S = S :=
+  eq_of_subset_of_card_le hσSS (ge_of_eq (card_image_of_injective S σ.to_ring_hom.injective)),
+  have hy : y ∈ σ '' S,
+  { rw [h3, polynomial.mem_root_set hp, minpoly.aeval] },
+  rcases hy with ⟨a, ha, hay⟩,
+  exact ⟨a, hay⟩,
+end
+
+lemma ultrafilter.glued_generators_of_pushforwards_alg_hom_bijection (h_int : algebra.is_integral
+  K L) (f : ultrafilter (L →ₐ[K] L)) : function.bijective
+  (ultrafilter.glued_generators_of_pushforwards_alg_hom f h_int) :=
+⟨ultrafilter.glued_generators_of_pushforwards_alg_hom_injective h_int f,
+  ultrafilter.glued_generators_of_pushforwards_alg_hom_surjective h_int f⟩
+
+/-- Since `ultrafilter.glued_generators_of_pushforwards_alg_hom h_int f` is a bijective `K`-algebra
+  homomorphism `L →ₐ[K] L`,it is a `K`-algebra equivalence. Here we define the corresponding term
+  of `L ≃ₐ[K] L`. -/
+noncomputable def ultrafilter.glued_generators_of_pushforwards_alg_equiv (h_int :
+  algebra.is_integral K L) (f : ultrafilter (L →ₐ[K] L)) : (L ≃ₐ[K] L) :=
+alg_equiv.of_bijective (ultrafilter.glued_generators_of_pushforwards_alg_hom f h_int)
+  (ultrafilter.glued_generators_of_pushforwards_alg_hom_bijection h_int f)
+
+/-- Let `L/K` be a normal algebraic field extension, let `f` be an ultrafilter on
+  `L ≃ₐ[K] L`, and let `E/K` be a finite subextension. Then
+  `ultrafilter.glued_generators_of_pushforwards_alg_equiv h_int f`
+  is a term of `L ≃ₐ[K] L`, and `alg_hom_of_finite_dimensional_of_ultrafilter h_findim f` is a term
+  `E →ₐ[K] L`. This Lemma tells us that the underlying maps of these two terms agree on `E`.
+    -/
+lemma equiv_restricts_to_alg_hom_of_finite_dimensional (h_int : algebra.is_integral K L)
+  (f : ultrafilter (L →ₐ[K] L)) {E : intermediate_field K L} (h_findim : finite_dimensional K E) :
+  ((ultrafilter.glued_generators_of_pushforwards_alg_equiv h_int f).to_alg_hom.comp E.val) =
+  f.generator_of_pushforward h_findim :=
+alg_hom.ext (ultrafilter.glued_generators_of_pushforwards_function_spec' h_int f h_findim)
+
+lemma inv_mul_alg_equiv_of_elem (x : L) (f g : L ≃ₐ[K] L) : (f⁻¹ * g) x = x ↔ g x = f x :=
+f.to_equiv.symm_apply_eq
+
+
+lemma top_group_map_nhds_eq {G : Type*} [group G] [topological_space G] [topological_group G]
+  (g x : G) : filter.map (λ y, g * y) (nhds x) = nhds (g * x) :=
+map_mul_left_nhds g x
+
```

Comment:
You could consider deleting these lemmas since they are no longer used and are already in mathlib.
```suggestion
```